# Questions

1) State various ways to handle missing values in the dataset. <br>
2) Apply Dropping Missing values procedure.<br>
3) Apply Imputation Procedure. <br>
4) Develop a procedure to assess both procedures (Hint: Use a regressor like randomforest Regressor, with a metric like MAE, and apply the dataset after each procedure to it) <br>
5) Give the values of the metrics you chose, also state whether imputation was a better method of Dropping missing values.<br>

In [61]:
# Import libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

from sklearn.model_selection import train_test_split

In [3]:
#Read the file
df = pd.read_csv("./6501242_IP02S4RN.csv")

In [4]:
df.head()

Suburb           Address  Rooms Type      Price Method SellerG  \
0  Abbotsford      85 Turner St      2    h  1480000.0      S  Biggin   
1  Abbotsford   25 Bloomburg St      2    h  1035000.0      S  Biggin   
2  Abbotsford      5 Charles St      3    h  1465000.0     SP  Biggin   
3  Abbotsford  40 Federation La      3    h   850000.0     PI  Biggin   
4  Abbotsford       55a Park St      4    h  1600000.0     VB  Nelson   

        Date  Distance  Postcode      ...       Bathroom  Car  Landsize  \
0  3/12/2016       2.5    3067.0      ...            1.0  1.0     202.0   
1  4/02/2016       2.5    3067.0      ...            1.0  0.0     156.0   
2  4/03/2017       2.5    3067.0      ...            2.0  0.0     134.0   
3  4/03/2017       2.5    3067.0      ...            2.0  1.0      94.0   
4  4/06/2016       2.5    3067.0      ...            1.0  2.0     120.0   

   BuildingArea  YearBuilt  CouncilArea Lattitude  Longtitude  \
0           NaN        NaN        Yarra  -37.7996    144.9984   
1          79.0     1900.0        Yarra  -37.8079    144.9934   
2         150.0     1900.0        Yarra  -37.8093    144.9944   
3           NaN        NaN        Yarra  -37.7969    144.9969   
4         142.0     2014.0        Yarra  -37.8072    144.9941   

              Regionname Propertycount  
0  Northern Metropolitan        4019.0  
1  Northern Metropolitan        4019.0  
2  Northern Metropolitan        4019.0  
3  Northern Metropolitan        4019.0  
4  Northern Metropolitan        4019.0  

[5 rows x 21 columns]

### 1) State various ways to handle missing values in the dataset. 

##### Dropping missing value
##### Imputation Using (Mean/Median) Values:
This works by calculating the mean/median of the non-missing values in a column and then replacing the missing values within each column separately and independently from the others. It can only be used with numeric data.

##### Imputation Using (Most Frequent) or (Zero/Constant) Values:
strategy to impute missing values and YES!! It works with categorical features (strings or numerical representations) by replacing missing data with the most frequent values within each column.
##### Imputation Using k-NN:
##### Imputation Using Multivariate Imputation by Chained Equation (MICE)

### 2) Apply Dropping Missing values procedure.

In [5]:
df.shape

(13580, 21)

In [45]:
df_droppped_missing_values = df.dropna(axis=0)

In [46]:
df_droppped_missing_values.shape

(6196, 21)

### 3) Apply Imputation Procedure. 

In [11]:
df.head()

Suburb           Address  Rooms Type      Price Method SellerG  \
0  Abbotsford      85 Turner St      2    h  1480000.0      S  Biggin   
1  Abbotsford   25 Bloomburg St      2    h  1035000.0      S  Biggin   
2  Abbotsford      5 Charles St      3    h  1465000.0     SP  Biggin   
3  Abbotsford  40 Federation La      3    h   850000.0     PI  Biggin   
4  Abbotsford       55a Park St      4    h  1600000.0     VB  Nelson   

        Date  Distance  Postcode      ...       Bathroom  Car  Landsize  \
0  3/12/2016       2.5    3067.0      ...            1.0  1.0     202.0   
1  4/02/2016       2.5    3067.0      ...            1.0  0.0     156.0   
2  4/03/2017       2.5    3067.0      ...            2.0  0.0     134.0   
3  4/03/2017       2.5    3067.0      ...            2.0  1.0      94.0   
4  4/06/2016       2.5    3067.0      ...            1.0  2.0     120.0   

   BuildingArea  YearBuilt  CouncilArea Lattitude  Longtitude  \
0           NaN        NaN        Yarra  -37.7996    144.9984   
1          79.0     1900.0        Yarra  -37.8079    144.9934   
2         150.0     1900.0        Yarra  -37.8093    144.9944   
3           NaN        NaN        Yarra  -37.7969    144.9969   
4         142.0     2014.0        Yarra  -37.8072    144.9941   

              Regionname Propertycount  
0  Northern Metropolitan        4019.0  
1  Northern Metropolitan        4019.0  
2  Northern Metropolitan        4019.0  
3  Northern Metropolitan        4019.0  
4  Northern Metropolitan        4019.0  

[5 rows x 21 columns]

In [13]:
#Shows null values
df.isnull().sum()

Suburb              0
Address             0
Rooms               0
Type                0
Price               0
Method              0
SellerG             0
Date                0
Distance            0
Postcode            0
Bedroom2            0
Bathroom            0
Car                62
Landsize            0
BuildingArea     6450
YearBuilt        5375
CouncilArea      1369
Lattitude           0
Longtitude          0
Regionname          0
Propertycount       0
dtype: int64

#### Imputing Car values

In [43]:
#Imputing Car values with most frequent value
df['Car'].value_counts().index[0]

2.0

In [44]:
df['Car'] = df['Car'].fillna(df['Car'].value_counts().index[0])

#### Imputing Building Area

In [49]:
df["BuildingArea"] = df["BuildingArea"].fillna(df['BuildingArea'].mean())

#### Imputing YearBuilt

In [50]:
df['YearBuilt'] = df['YearBuilt'].fillna(df['YearBuilt'].value_counts().index[0])

#### Imputing CouncilArea

In [53]:
df['CouncilArea'] = df['CouncilArea'].fillna(df['CouncilArea'].value_counts().index[0])

In [54]:
df.isnull().sum()

Suburb           0
Address          0
Rooms            0
Type             0
Price            0
Method           0
SellerG          0
Date             0
Distance         0
Postcode         0
Bedroom2         0
Bathroom         0
Car              0
Landsize         0
BuildingArea     0
YearBuilt        0
CouncilArea      0
Lattitude        0
Longtitude       0
Regionname       0
Propertycount    0
dtype: int64

### 4 ) Develop a procedure to assess both procedures (Hint: Use a regressor like randomforest Regressor, with a metric like MAE, and apply the dataset after each procedure to it) 

In [55]:
columns=['Car','BuildingArea','YearBuilt','CouncilArea']

#### 4.1 Test MAE for imputed data set

In [59]:
df.describe().columns

Index(['Rooms', 'Price', 'Distance', 'Postcode', 'Bedroom2', 'Bathroom', 'Car',
       'Landsize', 'BuildingArea', 'YearBuilt', 'Lattitude', 'Longtitude',
       'Propertycount'],
      dtype='object')

In [62]:
columns = ['Rooms', 'Distance', 'Postcode', 'Bedroom2', 'Bathroom', 'Car',
       'Landsize', 'BuildingArea', 'YearBuilt', 'Lattitude', 'Longtitude',
       'Propertycount']

In [102]:
features = df[columns]
labels = np.array(df['Price'])

In [105]:
# One-hot encode the data using pandas get_dummies
features = pd.get_dummies(features)
# Convert to numpy array
features = np.array(features)

In [96]:
# Split the data into training and testing sets
train_features, test_features, train_labels, test_labels = train_test_split(features, labels, test_size = 0.25, random_state = 42)

In [97]:
print('Training Features Shape:', train_features.shape)
print('Training Labels Shape:', train_labels.shape)
print('Testing Features Shape:', test_features.shape)
print('Testing Labels Shape:', test_labels.shape)

Training Features Shape: (10185, 12)
Training Labels Shape: (10185,)
Testing Features Shape: (3395, 12)
Testing Labels Shape: (3395,)


In [69]:
# Import the model we are using
from sklearn.ensemble import RandomForestRegressor

In [98]:
# Instantiate model with 250 decision trees
rf = RandomForestRegressor(n_estimators = 500, random_state = 42)
# Train the model on training data
rf.fit(train_features, train_labels);

In [99]:
# Use the forest's predict method on the test data
predictions = rf.predict(test_features)
# Calculate the absolute errors
errors = abs(predictions - test_labels)
# Print out the mean absolute error (mae)
print('Mean Absolute Error:', round(np.mean(errors), 2), 'degrees.')

Mean Absolute Error: 172966.92 degrees.


In [100]:
# Calculate mean absolute percentage error (MAPE)
mape = 100 * (errors / test_labels)
# Calculate and display accuracy
accuracy = 100 - np.mean(mape)
print('Accuracy:', round(accuracy, 2), '%.')

Accuracy: 83.46 %.


#### 4.2 Test MAE for dataframe with dropped missing values

In [73]:
features = df_droppped_missing_values [columns]
labels = df_droppped_missing_values ['Price']

In [74]:
# Split the data into training and testing sets
train_features, test_features, train_labels, test_labels = train_test_split(features, labels, test_size = 0.25, random_state = 42)

In [75]:
print('Training Features Shape:', train_features.shape)
print('Training Labels Shape:', train_labels.shape)
print('Testing Features Shape:', test_features.shape)
print('Testing Labels Shape:', test_labels.shape)

Training Features Shape: (4647, 12)
Training Labels Shape: (4647,)
Testing Features Shape: (1549, 12)
Testing Labels Shape: (1549,)


In [76]:
# Instantiate model with 1000 decision trees
rf = RandomForestRegressor(n_estimators = 500, random_state = 42)
# Train the model on training data
rf.fit(train_features, train_labels);

In [77]:
# Use the forest's predict method on the test data
predictions = rf.predict(test_features)
# Calculate the absolute errors
errors = abs(predictions - test_labels)
# Print out the mean absolute error (mae)
print('Mean Absolute Error:', round(np.mean(errors), 2), 'degrees.')

Mean Absolute Error: 166057.98 degrees.


In [79]:
# Calculate mean absolute percentage error (MAPE)
mape = 100 * (errors / test_labels)
# Calculate and display accuracy
accuracy = 100 - np.mean(mape)
print('Accuracy:', round(accuracy, 2), '%.')


Accuracy: 83.22 %.


### 5) Give the values of the metrics you chose, also state whether imputation was a better method of Dropping missing values.

Metric choosen is MAE. <br>
MAE for Imputed data frame is : 172966.92 <br>
MAE for dropped null value dtaa frame is : 166057.98 <br>


In this example, since MAE is less for the data frame where missing values were dropped, Dropping missing values is better. However mae might also depend on features selected and parameter tuning for model builidng